functions_framework_tool/test/validate_templates.dart

```diff
@@ -0,0 +1,206 @@
+
+final String _expectedGitIgnore = _getMetaTemplateFile('.gitignore');
+final String _expectedAnalysisOptions =
+    _getMetaTemplateFile('templates/analysis_options.yaml');
+final String _expectedAngularAnalysisOptions = [
+  _expectedAnalysisOptions.split('\n').take(12),
+  '  exclude: [build/**]',
+  '  errors:',
+  '    uri_has_not_been_generated: ignore',
+  "  # Angular plugin support is in beta. You're welcome to try it and report",
+  '  # issues: https://github.com/dart-lang/angular_analyzer_plugin/issues',
+  '  # plugins:',
+  '    # - angular\n',
+].expand((e) => e is Iterable ? e : [e]).join('\n');
+
+void main() {
+  Directory dir;
+
+  setUp(() async {
+    dir = await Directory.systemTemp.createTemp('stagehand.test.');
+  });
+
+  tearDown(() async {
+    if (dir != null) {
+      await dir.delete(recursive: true);
+    }
+  });
+
+  test('Meta-template .gitignore exists',
+      () => expect(_expectedGitIgnore, isNotEmpty));
+
+  test('Meta-template analysis_options.yaml exists',
+      () => expect(_expectedAnalysisOptions, isNotEmpty));
+
+  test('Validate pkg/stagehand pubspec', () {
+    var pubspecContent =
+        File(path.join(path.current, 'pubspec.yaml')).readAsStringSync();
+    _validatePubspec(pubspecContent);
+  });
+
+  group('generator', () {
+    for (var generator in stagehand.generators) {
+      test(generator.id, () {
+        _testGenerator(generator, dir);
+      });
+    }
+  });
+}
+
+void _testGenerator(stagehand.Generator generator, Directory tempDir) {
+  Dart.run(path.join(path.current, 'bin/dartfn.dart'),
+      arguments: ['--mock-analytics', generator.id],
+      runOptions: RunOptions(workingDirectory: tempDir.path));
+
+  var pubspecPath = path.join(tempDir.path, 'pubspec.yaml');
+  var pubspecFile = File(pubspecPath);
+  var pubspecContentString = pubspecFile.readAsStringSync();
+  var pubspecContent = yaml.loadYaml(pubspecContentString) as yaml.YamlMap;
+  final usesAngular =
+      pubspecContent['dependencies']?.containsKey('angular') ?? false;
+
+  var analysisOptionsPath = path.join(tempDir.path, 'analysis_options.yaml');
+  var analysisOptionsFile = File(analysisOptionsPath);
+  expect(
+    analysisOptionsFile.readAsStringSync(),
+    usesAngular ? _expectedAngularAnalysisOptions : _expectedAnalysisOptions,
+    reason: 'All analysis_options.yaml files should be identical.',
+  );
+
+  if (!pubspecFile.existsSync()) {
+    fail('A pubspec must be defined!');
+  }
+
+  Pub.get(runOptions: RunOptions(workingDirectory: tempDir.path));
+
+  var filePath = path.join(tempDir.path, generator.entrypoint.path);
+
+  if (path.extension(filePath) != '.dart' ||
+      !FileSystemEntity.isFileSync(filePath)) {
+    var parent = Directory(path.dirname(filePath));
+
+    var file = _listSync(parent)
+        .firstWhere((f) => f.path.endsWith('.dart'), orElse: () => null);
+
+    if (file == null) {
+      filePath = null;
+    } else {
+      filePath = file.path;
+    }
```

Comment:
Code like this can probably be put in a ternary.

```dart
filePath = (file == null) ? null : file.path;
```

Reply:
Yep, all of these test were copied directly from stagehand. I've got a ticket opened already for refactoring. I'll add this to the list of test improvements: #136

Reply:
Yup, tests come from stagehand ([here](https://github.com/dart-lang/stagehand/blob/master/test/validate_templates.dart#L112-L117)). See #136